src/array/array.cc

```diff
@@ -0,0 +1,372 @@
+  ATEN_XPU_SWITCH(arr->ctx.device_type, XPU, {
+    ATEN_ID_TYPE_SWITCH(arr->dtype, IdType, {
+      ret = impl::AsNumBits<XPU, IdType>(arr, bits);
+    });
+  });
+  return ret;
+}
+
+IdArray Add(IdArray lhs, IdArray rhs) {
+  IdArray ret;
+  CHECK_EQ(lhs->ctx, rhs->ctx) << "Both operands should have the same device context";
+  CHECK_EQ(lhs->dtype, rhs->dtype) << "Both operands should have the same dtype";
+  ATEN_XPU_SWITCH(lhs->ctx.device_type, XPU, {
+    ATEN_ID_TYPE_SWITCH(lhs->dtype, IdType, {
+      ret = impl::BinaryElewise<XPU, IdType, arith::Add>(lhs, rhs);
+    });
+  });
+  return ret;
+}
+
+IdArray Sub(IdArray lhs, IdArray rhs) {
+  IdArray ret;
+  CHECK_EQ(lhs->ctx, rhs->ctx) << "Both operands should have the same device context";
+  CHECK_EQ(lhs->dtype, rhs->dtype) << "Both operands should have the same dtype";
+  ATEN_XPU_SWITCH(lhs->ctx.device_type, XPU, {
+    ATEN_ID_TYPE_SWITCH(lhs->dtype, IdType, {
+      ret = impl::BinaryElewise<XPU, IdType, arith::Sub>(lhs, rhs);
+    });
+  });
+  return ret;
+}
+
+IdArray Mul(IdArray lhs, IdArray rhs) {
+  IdArray ret;
+  CHECK_EQ(lhs->ctx, rhs->ctx) << "Both operands should have the same device context";
+  CHECK_EQ(lhs->dtype, rhs->dtype) << "Both operands should have the same dtype";
+  ATEN_XPU_SWITCH(lhs->ctx.device_type, XPU, {
+    ATEN_ID_TYPE_SWITCH(lhs->dtype, IdType, {
+      ret = impl::BinaryElewise<XPU, IdType, arith::Mul>(lhs, rhs);
+    });
+  });
+  return ret;
+}
+
+IdArray Div(IdArray lhs, IdArray rhs) {
+  IdArray ret;
+  CHECK_EQ(lhs->ctx, rhs->ctx) << "Both operands should have the same device context";
+  CHECK_EQ(lhs->dtype, rhs->dtype) << "Both operands should have the same dtype";
+  ATEN_XPU_SWITCH(lhs->ctx.device_type, XPU, {
+    ATEN_ID_TYPE_SWITCH(lhs->dtype, IdType, {
+      ret = impl::BinaryElewise<XPU, IdType, arith::Div>(lhs, rhs);
+    });
+  });
+  return ret;
+}
+
+IdArray Add(IdArray lhs, dgl_id_t rhs) {
+  IdArray ret;
+  ATEN_XPU_SWITCH(lhs->ctx.device_type, XPU, {
+    ATEN_ID_TYPE_SWITCH(lhs->dtype, IdType, {
+      ret = impl::BinaryElewise<XPU, IdType, arith::Add>(lhs, rhs);
+    });
+  });
+  return ret;
+}
+
+IdArray Sub(IdArray lhs, dgl_id_t rhs) {
+  IdArray ret;
+  ATEN_XPU_SWITCH(lhs->ctx.device_type, XPU, {
+    ATEN_ID_TYPE_SWITCH(lhs->dtype, IdType, {
+      ret = impl::BinaryElewise<XPU, IdType, arith::Sub>(lhs, rhs);
+    });
+  });
+  return ret;
+}
+
+IdArray Mul(IdArray lhs, dgl_id_t rhs) {
+  IdArray ret;
+  ATEN_XPU_SWITCH(lhs->ctx.device_type, XPU, {
+    ATEN_ID_TYPE_SWITCH(lhs->dtype, IdType, {
+      ret = impl::BinaryElewise<XPU, IdType, arith::Mul>(lhs, rhs);
+    });
+  });
+  return ret;
+}
+
+IdArray Div(IdArray lhs, dgl_id_t rhs) {
+  IdArray ret;
+  ATEN_XPU_SWITCH(lhs->ctx.device_type, XPU, {
+    ATEN_ID_TYPE_SWITCH(lhs->dtype, IdType, {
+      ret = impl::BinaryElewise<XPU, IdType, arith::Div>(lhs, rhs);
+    });
+  });
+  return ret;
+}
+
+IdArray Add(dgl_id_t lhs, IdArray rhs) {
+  return Add(rhs, lhs);
```

Comment:
Why don't you do `impl::BinaryElewise<XPU, IdType, arith::Add>(lhs, rhs);` here?

Reply:
Because it's commutative, so I reuse the function to save several lines of codes.